src/callgraphwidget.cpp

```diff
@@ -0,0 +1,198 @@
+CallgraphWidget::CallgraphWidget(Data::CallerCalleeResults results, KParts::ReadOnlyPart* view, KGraphViewer::KGraphViewerInterface* interface, QWidget* parent)
+    : QWidget(parent)
+    , ui(new Ui::CallgraphWidget)
+    , m_graphFile(new QTemporaryFile(this))
+    , m_graphview(view)
+    , m_interface(interface)
+    , m_callerCalleeResults(results)
+{
+    ui->setupUi(this);
+
+    connect(ui->costThreshold, qOverload<double>(&QDoubleSpinBox::valueChanged), this, [this](double threshold) {
+        m_thresholdPercent = threshold / 100.f; // convert to percent
+        generateCallgraph(m_currentSymbol);
+    });
+
+    layout()->replaceWidget(ui->graphPlaceholder, view->widget());
+
+    updateColors();
+    m_interface->setLayoutMethod(KGraphViewer::KGraphViewerInterface::LayoutMethod::InternalLibrary);
+
+    connect(m_graphview, SIGNAL(hoverEnter(QString)), this, SLOT(hoverEnter(QString)));
+    connect(m_graphview, SIGNAL(hoverLeave(QString)), this, SLOT(hoverLeave(QString)));
+
+    // used to find mouseclicks in the kgraphviewer part
+    qApp->installEventFilter(this);
+
+    m_graphFile->open();
+
+    connect(Settings::instance(), &Settings::callgraphChanged, this, [this]{
+        generateCallgraph(m_currentSymbol);
+    });
+}
+
+CallgraphWidget::~CallgraphWidget() = default;
+
+CallgraphWidget *CallgraphWidget::createCallgraphWidget(const Data::CallerCalleeResults results, QWidget *parent)
+{
+    KService::Ptr service = KService::serviceByDesktopName(QStringLiteral("kgraphviewer_part"));
+
+    if (!service) {
+        return nullptr;
+    }
+
+    auto view = service->createInstance<KParts::ReadOnlyPart>(parent);
+    if (!view) {
+        return nullptr;
+    }
+
+    auto interface = qobject_cast<KGraphViewer::KGraphViewerInterface*>(view);
+    if (!interface) {
+        return nullptr;
+    }
+
+    return new CallgraphWidget(results, view, interface, parent);
+}
+
+void CallgraphWidget::selectSymbol(const Data::Symbol& symbol)
+{
+    if (symbol == m_currentSymbol) {
+        return;
+    }
+
+    generateCallgraph(symbol);
+}
+
+void CallgraphWidget::setResults(const Data::CallerCalleeResults &results)
+{
+    m_callerCalleeResults = results;
+    selectSymbol(m_currentSymbol);
+}
+
+void CallgraphWidget::hoverEnter(const QString& node)
+{
+    // cut of the node prefix
+    m_currentNode = node.mid(4);
+}
+
+void CallgraphWidget::hoverLeave(const QString& node)
+{
+    Q_UNUSED(node);
+    m_currentNode = QString();
+}
+
+bool CallgraphWidget::eventFilter(QObject* watched, QEvent* event)
+{
+    Q_UNUSED(watched);
+    if (!isVisible())
+        return false;
+
+    if (event->type() == QEvent::MouseButtonPress) {
+        const auto e = static_cast<QMouseEvent*>(event);
+
+        if (m_graphview->widget()->geometry().contains(e->pos())) {
+            if (e->button() == Qt::MouseButton::LeftButton && !m_currentNode.isEmpty()) {
+                for (auto it = m_symbolToId.cbegin(); it != m_symbolToId.cend(); it++) {
```

Comment:
this can be replaced by

```
const auto symbol = m_symbolToId.key(m_currentNode);
if (symbol.isValid()) {
    emit clickedOn(symbol);
    m_currentNode.clear();
    return true;
}
```